packages/react-hooks/src/useSubscription.js

```diff
@@ -0,0 +1,90 @@
+import {useEffect, useState} from 'react';
+
+export function useSubscription({
+  // This is the thing being subscribed to (e.g. an observable, event dispatcher, etc).
+  source,
+
+  // (Synchronously) returns the current value of our subscription source.
+  getCurrentValue,
+
+  // This function is passed an event handler to attach to the subscription source.
+  // It should return an unsubscribe function that removes the handler.
+  subscribe,
+}) {
+  // Read the current value from our subscription source.
+  // When this value changes, we'll schedule an update with React.
+  // It's important to also store the source itself so that we can check for staleness.
+  // (See the comment in checkForUpdates() below for more info.)
+  const [state, setState] = useState({
+    source,
+    value: getCurrentValue(source),
+  });
+
+  // If the source has changed since our last render, schedule an update with its current value.
+  if (state.source !== source) {
```

Comment:
isnt this somewhat a side effect in render phase? i suppose it leads to a predictable result even if a particular render gets replayed, but shouldnt generally this be done inside useEffect?

Reply:
No. A side effect would be e.g. mutating a variable or calling a callback. This is just telling React to schedule some follow up work.

Reply:
This is essentially following the pattern we recommend for derived state:
https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops

Reply:
Thanks for the clarification!

Reply:
I know you showed me this earlier, but I can't remember why this part is necessary given that `source` is one of the deps in the effect below.

Reply:
So the thing I'm guarding against here (which maybe my comment doesn't do a good job of clarifying) is this:

1. Subscription added to source A
1. Component renders with a new source, source B
1. Source A emits an update
1. Passive effect is invoked, removing subscription from A and adding to B

We need to ignore the update from A that happens after we get a new source (but before our passive effect is fired).

I tried to make this clear with all of the inline comments but maybe I could improve them somehow?

Reply:
Is `didUnsubscribe` not sufficient for that?

Reply:
No. Because in the case I mentioned above, we **haven't unsubscribed** yet (because the passive effect wasn't yet fired).

Reply:
We always fire pending passive effects right at the beginning of `setState`, to prevent this type of scenario

Reply:
Ah but I guess that would require closing over a mutable variable. Ok now I get it.